<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>UTHSC Hippocampus Illumina v6.1 RSS (Sep09) RankInv **</TITLE>
<META http-equiv=Content-Type content="text/html; charset=iso-8859-1">

<LINK REL="stylesheet" TYPE="text/css" HREF='/css/general.css'>
<LINK REL="stylesheet" TYPE="text/css" HREF='/css/menu.css'>
<link rel="stylesheet" media="all" type="text/css" href="/css/tabbed_pages.css" />
<SCRIPT SRC="/javascript/webqtl.js"></SCRIPT>

<SCRIPT SRC="/javascript/dhtml.js"></SCRIPT>

<script src="/javascript/tabbed_pages.js" type="text/javascript"></script>

</HEAD>
<BODY  bottommargin="2" leftmargin="2" rightmargin="2" topmargin="2" text=#000000 bgColor=#ffffff >

<TABLE cellSpacing=5 cellPadding=4 width="100%" border=0>
	<TBODY>
	<TR>
		<script language="JavaScript" src="/javascript/header.js"></script>
	</TR>
	<TR>
		<TD  bgColor=#eeeeee class="solidBorder">
		<Table width= "100%" cellSpacing=0 cellPadding=5>
		<TR>

		<!-- split from Here -->
		<!-- Body Start from Here -->
		<P class="title">UTHSC Illumina Whole Mouse Genome Array 6.0 Version 2 data set   <A HREF="/webqtl/main.py?FormID=editHtml">
					<img src="/images/modify.gif" alt="modify this page" border= 0 valign="middle"></A></P>
<P>Modified by A. Centeno, Feb 2, 2016
<P>Array data set generated by Dr. Xusheng Wang (2008)
<P><input type="button" name="Button" class="button" value="Download Dataset" onClick="window.open('ILM_B_D_CNS_GI_avg_0508.xls', 'download'); return false;"></P>
<p><a href="http://genenetwork.org/images/upload/MouseTissuesRWWMay08.xls">33 Mouse tissues expression values (excel file)</a>
<h3>Animals</h3>
<p>Two individual male strains (C57BL/6J and DBA/2J) were sacrifice to extract 61 tissues, including 32 CNS and 5 GIs (See tissue names in the table). The uterus of the C57BL/6J and DBA/2J mice were obtained from the corresponding female strains.  Around two month old C57BL/6J and DBA/2J mice were used.	
<h3>Tissue Collection and RNA isolation</h3>
<p>	The mouse was maintained at 20-24&deg;C on a 12/12 hr light/dark cycle in a pathogen-free colony at the University of Tennessee, and was fed a 5% fat Agway Prolab 3000 rat and mouse chow and given tap water in glass bottles. Mice were sacrificed by cervical dislocation. 27 different tissues and organs were dissected from the body and place in RNAlater and then stored at -80&deg;C until use. The whole brain was also rapidly extracted and placed on a sagital matrix with 1 mm divisions. The extraction of RNA from hippocampal tissues was carried out with the RNA STAT-60 reagent (Tel-Test Inc, Friedenswood, TX, USA). Tissue samples were washed three times in phosphate-buffered saline (PBS; GibCo BRL, Grand Island, NY, USA) to remove blood contamination. 100mg tissue was homogenized in 1 ml of RNA STAT-60 reagent. Following homogenization, store the homogenate for 5 minutes at room temperature. Next, 0.2ml of chloroform per 1ml of RNA STAT-60 was added and the mix was vigorously shaken for 15 seconds and centrifuged at 12,000g for 15 minutes at 4&deg;C. After centrifugation the aqueous phase was transferred to a fresh tube and mixed with 0.5ml of isopropanol per 1ml of RNA STAT-60 used for the homogenization. The precipitate was washed twice in 75% ethanol, air-dried and re-diluted in Nuclease-free water (Ambion Inc., TX, USA). The purity of the extracted RNA was determined by by NanoDrop spectrophotometer (NanoDrop Technologies Inc, NC, USA). The 260/280 ratio of the samples was within the desired range of 1.9-2.1. RNA integrity was assessed on the Agilent 2100 Bioanalyzer (Agilent Technologies, CA, USA), and the RNA Integrity Number (RIN) values were required greater than 8.0.
<h3>Bead array and gene expression analysis</h3>
<p>Sample amplification was performed with 100 ng of total RNA using the Illumina TotalPrep RNA Amplification kit and labeling was achieved by incorporation of biotin-16-UTP (Perkin Elmer Life and Analytical Sciences) at a ratio of 1:1 with unlabeled UTP. Labeled, amplified material (100 ng per array) was hybridized to Illumina Bead chips according to the Manufacturer's instructions (Illumina, Inc.). Arrays were scanned with an Illumina Bead Array Reader confocal scanner according to the Manufacturer's instructions. Array data processes were performed using Illumina BeadStudio software. For the striatum of the BXD RI data, Illumina MouseWG-6v1 presented with 45,281 transcripts were used; for the tissue data, Illumina MouseWG-6V2 presented with 45,281 transcripts were used.

<p>To associate probes with RefSeq transcripts, we mapped the probes back to the genomes (NCBI mouse genome assembly m36) to identify the probe locations and exon targets. We used the resulting probe-to-exon map to identify the RefSeq transcripts targeted by each probe, and assign a probe set to each transcript.

<p>Normalization was performed by the rank variance method using Beadstudio software. We generated probe set data using Rank variance, obtained the log2 of each probe set and standardized using Z scores. We doubled the Z scores and added 8 to produce a set of Z scores with a mean of 8, a variance of 4 and a standard deviation of 2. The advantage of this modified Z score is that a two fold difference in expression level corresponds approximately to a 1-unit difference. Expression levels below 6 are usually close to background noise levels. 			
		</TR></TABLE>
		</TD>
	</TR>
	<TR>
		<TD align=center bgColor=#ddddff class="solidBorder">
		<!--Start of footer-->
		<TABLE width="90%">
			<script language='JavaScript' src='/javascript/footer.js'></script>
			<TR>
				<TD colspan=3 class="fs12">
					<UL>
					
					</UL>
				</TD>
			</TR>
		</TABLE>
		<!--End of footer-->
		</TD>
	</TR>
</TABLE>
<!-- /Footer -->
<!-- menu script itself. you should not modify this file -->
<script language="JavaScript" src="/javascript/menu_new.js"></script>
<!-- items structure. menu hierarchy and links are stored there -->
<script language="JavaScript" src="/javascript/menu_items.js"></script>
<!-- files with geometry and styles structures -->
<script language="JavaScript" src="/javascript/menu_tpl.js"></script>
<script language="JavaScript">
	<!--//
	// Note where menu initialization block is located in HTML document.
	// Don't try to position menu locating menu initialization block in
	// some table cell or other HTML element. Always put it before </body>
	// each menu gets two parameters (see demo files)
	// 1. items structure
	// 2. geometry structure
	new menu (MENU_ITEMS, MENU_POS);
	// make sure files containing definitions for these variables are linked to the document
	// if you got some javascript error like "MENU_POS is not defined", then you've made syntax
	// error in menu_tpl.js file or that file isn't linked properly.
	
	// also take a look at stylesheets loaded in header in order to set styles
	//-->

</script>
</BODY>
</HTML>